datarepo-clienttests/src/main/java/deploymentscripts/ModularHelmChart.java

```diff
@@ -0,0 +1,348 @@
+    }
+
+    // list the available deployments (for debugging)
+    // e.g. helm ls --namespace mm
+    ArrayList<String> listCmdArgs = new ArrayList<>();
+    listCmdArgs.add("ls");
+    listCmdArgs.add("--namespace");
+    listCmdArgs.add(serverSpecification.namespace);
+    Process helmListProc = ProcessUtils.executeCommand("helm", listCmdArgs);
+    cmdOutputLines = ProcessUtils.waitForTerminateAndReadStdout(helmListProc);
+    for (String cmdOutputLine : cmdOutputLines) {
+      System.out.println(cmdOutputLine);
+    }
+
+    // install/upgrade the API deployment using the modified YAML file we just generated
+    // e.g. helm namespace upgrade mm-jade-datarepo-api datarepo-helm/datarepo-api --install
+    // --namespace mm -f datarepo-api_MODIFIED.yaml
+    ArrayList<String> installCmdArgs = new ArrayList<>();
+    installCmdArgs.add("namespace");
+    installCmdArgs.add("upgrade");
+    installCmdArgs.add(serverSpecification.namespace + "-jade-datarepo-api");
+    installCmdArgs.add("datarepo-helm/datarepo-api");
+    installCmdArgs.add("--install");
+    installCmdArgs.add("--namespace");
+    installCmdArgs.add(serverSpecification.namespace);
+    installCmdArgs.add("-f");
+    installCmdArgs.add(modifiedApiYamlFile.getAbsolutePath());
+    Process helmUpgradeProc = ProcessUtils.executeCommand("helm", installCmdArgs);
+    cmdOutputLines = ProcessUtils.waitForTerminateAndReadStdout(helmUpgradeProc);
+    for (String cmdOutputLine : cmdOutputLines) {
+      System.out.println(cmdOutputLine);
+    }
+
+    // delete the two temp YAML files created above
+    boolean originalYamlFileDeleted = originalApiYamlFile.delete();
+    if (!originalYamlFileDeleted) {
+      throw new RuntimeException(
+          "Error deleting the _ORIGINAL YAML file: " + originalApiYamlFile.getAbsolutePath());
+    }
+    boolean modifiedYamlFileDeleted = modifiedApiYamlFile.delete();
+    if (!modifiedYamlFileDeleted) {
+      throw new RuntimeException(
+          "Error deleting the _MODIFIED YAML file: " + modifiedApiYamlFile.getAbsolutePath());
+    }
+  }
+
+  /**
+   * 1. Poll the deployment status with Helm until it reports that the datarepo-api is "deployed".
+   * 2. Poll the unauthenticated status endpoint until it returns success.
+   *
+   * <p>Waiting for the deployment to be ready to respond to API requests, often takes a long time
+   * (~5-15 minutes) because we deleted the deployment before re-installing it. This restarts the
+   * oidc-proxy which is what's holding things up. If we skip the delete deployment and just
+   * re-install, then this method usually returns much quicker (~1-5 minutes). We need to do the
+   * delete in order for the databases (Stairway and Data Repo) to be wiped because of how they
+   * decide which pod will do the database migration.
+   */
+  public void waitForDeployToFinish() throws Exception {
+    int pollCtr = Math.floorDiv(maximumSecondsToWaitForDeploy, secondsIntervalToPollForDeploy);
```

Comment:
I saw we're also doing something like this in the DataRepoUtils.waitForJobToFinish. Do you think a scaled backoff for the wait time would be helpful in either of these waits? (i.e. start with a small wait time and gradually increase it)

Reply:
I like the idea of a scaled backoff. I'll add a new ticket for implementing it in both places.